control-plane/agents/src/eventing/volume_events.rs

```diff
@@ -0,0 +1,25 @@
+use events_api::event::{EventAction, EventCategory, EventMessage, EventMeta, EventSource};
+use stor_port::types::v0::transport::VolumeState;
+
+use super::{EventMetaGen, EventWithMeta};
+
+// Volume state change event
+impl EventWithMeta for VolumeState {
+    fn event(&self, event_action: EventAction, meta: EventMeta) -> EventMessage {
+        EventMessage {
+            category: EventCategory::Volume as i32,
+            action: event_action as i32,
+            target: self.uuid.to_string(),
+            metadata: Some(meta),
+        }
+    }
+}
+
+// Get event meta data for volume state change events event
+impl EventMetaGen for VolumeState {
+    fn meta(&self) -> EventMeta {
+        let event_source =
+            EventSource::new("".to_string()).with_state_change_data("", &self.status.to_string());
```

Comment:
What are these "" ?

Reply:
First one is node, but couldn't find the node name here. 
Second one is previous state for volume state change. I added state change event for volume only when volume is getting degraded and also was not able to find the previous state. 

I couldn't add the rest of the state change events for volume and pools.

Reply:
Yeah I guess node doesn't make sense here? Or should it be the nexus node, if any?
But I don't think we should issue degraded volume state change tbh, we're missing the infra to have this in place.

Reply:
Yes, I removed it.